### ISSUE TITLE
Shared cache writes 0 byte keyfiles

### ISSUE THREAD

**Issue by adamse** [ndmitchell/shake#757]
We have observed 0 byte long key files in the shared cache, they cause terrible problems when trying to read the cached files back:

```
  at error, called at src/General/Binary.hs:54:112 in shake-0.19-5VUKQTr8rH42mS8NIVziBX:General.Binary

* Raised the exception:

Reading from ByteString, insufficient left
```

A possible problem is that [`Development.Shake.Internal.History.Shared.saveSharedEntry`][1] does not write the `.../_key/<hash>` files atomically.

I think we first observed this problem when we upgraded to GHC 8.10, but the non-atomicity of the key creation is not new.

[1]: https://github.com/ndmitchell/shake/blob/master/src/Development/Shake/Internal/History/Shared.hs#L140

**Comment by pepeiborra** [comment]
http://hackage.haskell.org/package/atomic-write could help perhaps? In ghcide we [roll](https://github.com/digital-asset/ghcide/blob/master/src/Development/IDE/Core/Compile.hs#L279) our own solution, but I cannot remember why

**Comment by ndmitchell** [comment]
Thanks for the report! I rolled my own too, since it avoids a dependency that pulls in text, and was only an extra line at that point in the code. Please reraise if you think I screwed it up - planning a release of Shake tomorrow which will include it.

**Comment by ndmitchell** [comment]
I released 0.19.1 which has this fix in it.
